File: frontend/src/functions/http-client/api/source-data.ts
import { AlreadyExists, NotFound, ResourceNotExist } from "../exceptions"
import { HttpInstance, Response } from ".."
import { LimitAndOffsetFilter, ListResult, mapFromOrderList, OrderList } from "./all"
import { SimpleIllust } from "./illust"
import { datetime, LocalDateTime } from "@/utils/datetime"

export function createSourceDataEndpoint(http: HttpInstance): SourceDataEndpoint {
    return {
        list: http.createQueryRequest("/api/source-data", "GET", {
            parseQuery: mapFromSourceDataFilter,
            parseResponse: ({ total, result }: ListResult<any>) => ({total, result: result.map(mapToSourceData)})
        }),
        create: http.createDataRequest("/api/source-data", "POST", {
            parseData: mapFromSourceDataUpdateForm
        }),
        bulk: http.createDataRequest("/api/source-data/bulk", "POST"),
        get: http.createPathRequest(({ sourceSite, sourceId }) => `/api/source-data/${encodeURIComponent(sourceSite)}/${encodeURIComponent(sourceId)}`, "GET", {
            parseResponse: mapToDetailSourceData
        }),
        update: http.createPathDataRequest(({ sourceSite, sourceId }) => `/api/source-data/${encodeURIComponent(sourceSite)}/${encodeURIComponent(sourceId)}`, "PATCH", {
            parseData: mapFromSourceDataUpdateForm
        }),
        delete: http.createPathRequest(({ sourceSite, sourceId }) => `/api/source-data/${encodeURIComponent(sourceSite)}/${encodeURIComponent(sourceId)}`, "DELETE"),
        getRelatedImages: http.createPathRequest(({ sourceSite, sourceId }) => `/api/source-data/${encodeURIComponent(sourceSite)}/${encodeURIComponent(sourceId)}/related-images`)
    }
}

function mapToSourceData(data: any): SourceData {
    return {
        ...data,
        publishTime: data.publishTime ? datetime.of(<string>data["publishTime"]) : null,
        createTime: datetime.of(<string>data["createTime"]),
        updateTime: datetime.of(<string>data["updateTime"]),
    }
}

function mapToDetailSourceData(data: any): DetailSourceData {
    return {
        ...data,
        publishTime: data.publishTime ? datetime.of(<string>data["publishTime"]) : null,
        createTime: datetime.of(<string>data["createTime"]),
        updateTime: datetime.of(<string>data["updateTime"]),
    }
}

function mapFromSourceDataUpdateForm(form: SourceDataUpdateForm): any {
    return {
        ...form,
        publishTime: form.publishTime && datetime.toISOString(form.publishTime),
    }
}

function mapFromSourceDataFilter(filter: SourceDataFilter): any {
    return {
        ...filter,
        order: mapFromOrderList(filter.order)
    }
}

export interface SourceDataEndpoint {
    list(filter: SourceDataFilter): Promise<Response<ListResult<SourceData>>>
    create(form: SourceDataCreateForm): Promise<Response<null, SourceDataExceptions["create"]>>
    bulk(items: SourceDataCreateForm[]): Promise<Response<null, SourceDataExceptions["bulk"]>>
    get(key: SourceDataIdentity): Promise<Response<DetailSourceData, NotFound>>
    getRelatedImages(key: SourceDataIdentity): Promise<Response<SimpleIllust[]>>
    update(key: SourceDataIdentity, form: SourceDataUpdateForm): Promise<Response<null, SourceDataExceptions["update"]>>
    delete(key: SourceDataIdentity): Promise<Response<null, NotFound>>
}

export interface SourceDataIdentity { sourceSite: string, sourceId: string }

export interface SourceDataExceptions {
    "create": ResourceNotExist<"site", string> | ResourceNotExist<"additionalInfo", string> | ResourceNotExist<"sourceTagType", string[]> | AlreadyExists<"SourceData", "sourceId", number>
    "bulk": ResourceNotExist<"site", string> | ResourceNotExist<"additionalInfo", string> | ResourceNotExist<"sourceTagType", string[]>
    "update": ResourceNotExist<"site", string> | ResourceNotExist<"additionalInfo", string> | ResourceNotExist<"sourceTagType", string[]> | NotFound
}

export type SourceEditStatus = "NOT_EDITED" | "EDITED" | "ERROR" | "IGNORED"

interface BasicSourceData {
    /**
     * source name。
     */
    sourceSite: string
    /**
     * source标题。自动从setting取得并填充。
     */
    sourceSiteName: string
    /**
     * source id。
     */
    sourceId: string
    /**
     * 是否为空。
     */
    empty: boolean
    /**
     * 编辑状态。
     */
    status: SourceEditStatus
    /**
     * 发布时间。
     */
    publishTime: LocalDateTime | null
    /**
     * 创建时间。
     */
    createTime: LocalDateTime
    /**
     * 更改时间。
     */
    updateTime: LocalDateTime
}

export interface SourceData extends BasicSourceData {
    /**
     * tag数量。
     */
    tagCount: number
    /**
     * book数量。
     */
    bookCount: number
    /**
     * relation数量。
     */
    relationCount: number
}

export interface DetailSourceData extends BasicSourceData {
    /**
     * 标题。
     */
    title: string
    /**
     * 描述。
     */
    description: string
    /**
     * 标签。
     */
    tags: SourceTag[]
    /**
     * book。
     */
    books: SourceBook[]
    /**
     * 相关项。
     */
    relations: string[]
    /**
     * 相关链接。
     */
    links: string[]
    /**
     * 附加信息。
     */
    additionalInfo: SourceAdditionalInfo[]
}

export interface SourceTag {
    code: string
    type: string
    name: string
    otherName: string | null
}

export interface SourceBook {
    code: string
    title: string
    otherTitle: string | null
}

export interface SourceAdditionalInfo {
    field: string
    label: string
    value: string
}

export interface SourceDataCreateForm extends SourceDataUpdateForm {
    sourceSite: string
    sourceId: string
}

export interface SourceDataUpdateForm {
    status?: SourceEditStatus
    title?: string
    description?: string
    tags?: SourceTagForm[]
    books?: SourceBookForm[]
    relations?: string[]
    links?: string[]
    additionalInfo?: SourceAdditionalInfoForm[]
    publishTime?: LocalDateTime | null
}

export interface SourceTagForm {
    code: string
    type: string
    name?: string
    otherName?: string | null
}

export interface SourceBookForm {
    code: string
    title?: string
    otherTitle?: string | null
}

export interface SourceAdditionalInfoForm {
    field: string
    value: string
}

export type SourceDataFilter = SourceDataQueryFilter & LimitAndOffsetFilter

export interface SourceDataQueryFilter {
    /**
     * 使用HQL进行查询。list API不提示解析结果。
     */
    query?: string
    /**
     * 排序字段列表。
     */
    order?: OrderList<"rowId" | "sourceId" | "site" | "createTime" | "updateTime" | "publishTime">
    /**
     * 按status类型过滤。
     */
    status?: SourceEditStatus[]
    /**
     * 按source site类型过滤。
     */
    site?: string[]
    /**
     * 按source tag过滤。
     */
    sourceTag?: string
    /**
     * 按关联的image id过滤。
     */
    imageId?: number
}

File: server/src/main/kotlin/com/heerkirov/hedge/server/model/SourceData.kt
package com.heerkirov.hedge.server.model

import com.heerkirov.hedge.server.enums.MetaType
import com.heerkirov.hedge.server.enums.SourceEditStatus
import java.time.Instant

/**
 * 来源信息。
 * 分离存储illust关联的来源信息，与image类型的illust 1:1存储。
 */
data class SourceData(val id: Int,
                      /**
                        * 来源网站的代号。
                        */
                      val sourceSite: String,
                      /**
                        * 来源网站中的图像id。
                        */
                      val sourceId: String,
                      /**
                       * 可排序的图像id。
                       */
                      val sortableSourceId: Long?,
                      /**
                        * 原数据的标题信息，有些会有，比如pixiv。
                        */
                      val title: String? = null,
                      /**
                        * 原数据的描述信息，有些会有，比如pixiv。
                        */
                      val description: String? = null,
                      /**
                        * 原数据的关系信息。
                        */
                      val relations: List<String>? = null,
                      /**
                       * 原数据的额外追加信息。它的数据条目需要定义，在site中定义。
                       */
                      val additionalInfo: Map<String, String>? = null,
                      /**
                       * 此项的发布时间。
                       */
                      val publishTime: Instant? = null,
                      /**
                        * 关系信息的数量的缓存。
                        */
                      val cachedCount: SourceCount,
                      /**
                        * 内容是否为空的缓存标记。
                        */
                      val empty: Boolean,
                      /**
                        * 此项的编辑状态。能自动转换。
                        */
                      val status: SourceEditStatus,
                      /**
                        * 初次建立的真实时间。
                        */
                      val createTime: Instant,
                      /**
                        * 上次更新的真实更新时间。
                        */
                      val updateTime: Instant) {

    data class SourceCount(val tagCount: Int, val bookCount: Int, val relationCount: Int)

}

/**
 * 来源信息的book。
 */
data class SourceBook(val id: Int,
                      /**
                       * 来源网站的代号。
                       */
                      val site: String,
                      /**
                       * 编码。
                       */
                      val code: String,
                      /**
                       * 标题。
                       */
                      val title: String,
                      /**
                       * 其他标题。
                       */
                      val otherTitle: String?)

/**
 * source image与book的关联。
 */
data class SourceBookRelation(val sourceDataId: Int, val sourceBookId: Int)

/**
 * 来源信息的标签。
 */
data class SourceTag(val id: Int,
                     /**
                      * 来源网站的代号。
                      */
                     val site: String,
                     /**
                      * 标签分类。
                      */
                     val type: String,
                     /**
                      * 标签编码。
                      */
                     val code: String,
                     /**
                      * 标签的显示名称。
                      */
                     val name: String,
                     /**
                      * 标签的其他名称。
                      */
                     val otherName: String?)

/**
 * source image与tag的关联。
 */
data class SourceTagRelation(val sourceDataId: Int, val sourceTagId: Int)

/**
 * 原始标签映射。
 * 记录由来源信息的tag到app tag的映射关系，符合映射表的tag会放入建议列表。
 */
data class SourceTagMapping(val id: Int,
                            /**
                             * 来源网站的代号。
                             */
                            val sourceSite: String,
                            /**
                             * 来源tag id。
                             */
                            val sourceTagId: Int,
                            /**
                             * 转换为什么类型的tag。
                             */
                            val targetMetaType: MetaType,
                            /**
                             * 目标tag的tag id。
                             */
                            val targetMetaId: Int)
